# Differential expression analysis

### Report generated by `r labkey.user.email`


This report creates list of differentially expressed genes comparing all available timepoints to baseline. 


```{r knitr-options, echo=FALSE, eval=TRUE, cache=FALSE}
library(knitr)
opts_chunk$set(echo=TRUE, eval=TRUE, message=FALSE)
opts_chunk$set(cache=FALSE, cache.path=paste(labkey.file.root, "cache/limma_analysis/", sep="/"))
options(width=80)
```

```{r load-required-packages, cache=FALSE}
library(Biobase)
library(limma)
library(Rlabkey)
library(data.table)
library(reshape2)
```

### Get the expression matrices from the database
```{r expression, cache=TRUE}
exMat <- data.table(labkey.selectRows(baseUrl = labkey.url.base, folderPath = labkey.url.path, schemaName = "study", queryName = "gene_expression_matrices", colNameOpt="rname"))
```

### Get the probe mapping file
```{r f2g, cache=TRUE, dependson="expression"}
fdLink <- paste(labkey.file.root, "analysis/features2genes", unique(exMat$feature_mapping_file), sep="/")
fdata <- read.table(fdLink, header=TRUE)
```
### Create ExpressionSets
```{r esets, cache=TRUE, dependson="expression"}
matrices <- unique(exMat$expression_matrix_accession)
nMatrices <- length(matrices)
eset_list <- vector('list', nMatrices)

for(idx in 1:length(matrices)){
  mat <- matrices[idx]
  pdata <- exMat[expression_matrix_accession == mat]
  pdata <- pdata[study_time_reported < 0, study_time_reported := "0"]
  em_link <- paste(labkey.file.root, "analysis/exprs_matrices", unique(pdata$file_info_name), sep="/")
  exprs <- fread(em_link)
  rn <- exprs$feature_id
  exprs <- exprs[, colnames(exprs) %in% pdata$biosample_accession, with=FALSE]
  exprs <- as.matrix(exprs)
  rownames(exprs) <- rn
  pdata <- pdata[na.omit(match(colnames(exprs), pdata$biosample_accession)),]
  pdata <- data.frame(pdata)
  rownames(pdata) <- pdata$biosample_accession
  rownames(fdata) <- fdata$feature_id
  eset <- ExpressionSet(assayData = as.matrix(exprs), phenoData = as(pdata, "AnnotatedDataFrame"), featureData = as(fdata, "AnnotatedDataFrame"))
  eset_list[[idx]] <- eset
}
```

### Differential expression analysis using limma
```{r limmma, cache=TRUE, depensdon="esets"}
FDRthresh <- 0.05
GEAcount <- 10

analysis_list <- vector("list", nMatrices)
analysis_res <- vector("list", nMatrices)

for(esetIdx in 1:nMatrices){
eset <- eset_list[[esetIdx]]

mm <- model.matrix(~subject_accession + study_time_reported, eset)
fit <- lmFit(eset, mm)
ebay <- eBayes(fit)

coefs <- grep("study_time_reported", colnames(mm), value=TRUE)
analysis_list[[esetIdx]] <- vector("list", length(coefs))
analysis_res[[esetIdx]] <- vector("list", length(coefs))
for(coefIdx in 1:length(coefs)){
   tt <- data.table(topTable(ebay, coef=coefs[coefIdx], number=Inf))
   AA <- paste0("GEA", GEAcount)
   tt <- tt[, analysis_accession := AA]
   GEAcount <- GEAcount + 1
   analysis_list[[esetIdx]][[coefIdx]] <- data.table(analysis_accession = AA, expression_matrix_accession = matrices[esetIdx], description = paste0("Differential expression analysis using limma D", gsub("study_time_reported", "", coefs[coefIdx]), " vs baseline"))
   analysis_res[[esetIdx]][[coefIdx]] <- tt
}
analysis_list[[esetIdx]] <- rbindlist(analysis_list[[esetIdx]])
analysis_res[[esetIdx]] <- rbindlist(analysis_res[[esetIdx]])[, list(analysis_accession, feature_id, adj.P.Val, AveExpr, logFC, P.Value, t)]
setnames(analysis_res[[esetIdx]], c("adj.P.Val", "AveExpr", "logFC", "P.Value", "t"), c("adj_p_val", "ave_expr", "log_fc", "p_value", "statistic"))
}
analysis_list <- rbindlist(analysis_list)
analysis_res <- rbindlist(analysis_res)
```

### Get existing analysis
```{r query-GEA-list, warning=FALSE}
analysis_filter <- makeFilter(c("analysis_accession", "IN", paste(analysis_list$analysis_accession, collapse=";")))
existing_analysis <- labkey.selectRows(labkey.url.base, labkey.url.path, "lists", "gene_expression_analysis", colFilter=analysis_filter, colNameOpt="rname")
```

### save results in tables
```{r write-lists, eval=FALSE}
toImport_GEA <- analysis_list[!(analysis_accession %in% existing_analysis$analysis_accession)]
if(nrow(toImport_GEA) > 0 ){
   imported_GEA <- labkey.importRows(labkey.url.base, labkey.url.path, "lists", "gene_expression_analysis", toImport = toImport_GEA)
}
#toImport <- analysis_res
write.table(toImport, file = file.path(labkey.file.root, "analysis/gene_expression_analysis/limma", "all_GEA.tsv"), sep="\t", quote=FALSE, row.names=FALSE)
#imported_GEARW <- labkey.importRows(labkey.url.base, labkey.url.path, "lists", "gene_expression_analysis_result_wide", toImport = toImport)
```
`r nrow(toImport_GEA)` new gene expression analysis saved.
